js/ddmlib/tl-worker.js

```diff
@@ -0,0 +1,54 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+importScripts("property_formatter.js")
+importScripts("../../third_party/protobuf.min.js")
+importScripts("viewnode.js")
+
+let rootNodes /* ViewNode[] */
+let formattingIndex = 0
+
+self.onmessage = function(event) {
+    if (rootNodes == null) {
+        protobuf.load("../../protos/view_capture.proto").then(function(root) {
+            rootNodes = root.lookupType("com.android.launcher3.view.ExportedData")
+                            .decode(event.data.tlHvDataAsBinaryArray)
+                            .frameData
+                            .map(f => f.node)
+            postMessage({ frameCount: rootNodes.length })
+            formatNextNode()
+        })
+    } else {
+        sendNodeToUiThread(event.data.processedIndex)
+    }
+}
+
+/* If the Ui thread processes nodes faster than the worker thread can format them,
+   wait for the next node to be formatted before sending it. */
+function sendNodeToUiThread(processedIndex) {
+    if (processedIndex < formattingIndex && processedIndex < rootNodes.length) {
```

Comment:
I am not sure we need incremental update here. The proto was already parsed as a full tree so we have essentially gone through all the nodes once, going it again should not add more delay.

Do you have an estimate of how long this takes if you format all the nodes in one go?

Reply:
To format 1 view hierarchy on the background thread, it usually takes between 0.5 and 1.3ms. Also, unfortunately it takes ~550ms to copy all 170 view hierarchies from the background thread to the main thread after they have been parsed.

Both of these latency additions make noticeable differences to the UX. With this incremental optimization, the time lapse UX is as fast and jank free as the single view hierarchy UX. Without it there more than an additional second of  latency before the user can interact with the webpage.